javalib/src/main/scala/java/util/concurrent/LinkedTransferQueue.scala

```diff
@@ -0,0 +1,1804 @@
+            }) {
+          if (h == null) h = head
+          if (p.tryMatch(item, e)) {
+            if (h != p) skipDeadNodesNearHead(h, p)
+            return item.asInstanceOf[E]
+          }
+        }
+        val q = p.next
+        if (q == null) {
+          if (how == NOW) return e
+          if (s == null) s = new Node(e)
+          if (!p.casNext(null, s)) skipRest = true
+          if (!skipRest) {
+            if (p != t) casTail(t, s)
+            if (how == ASYNC) return e
+            return awaitMatch(s, p, e, (how == TIMED), nanos)
+          }
+        }
+        if (!skipRest) {
+          val old_p = p
+          p = q
+          if (old_p == p) innerBreak = true
+        }
+      }
+    }
+    ???
+  }
+
+  /** Possibly blocks until node s is matched or caller gives up.
+   *
+   *  @param s
+   *    the waiting node
+   *  @param pred
+   *    the predecessor of s, or null if unknown (the null case does not occur
+   *    in any current calls but may in possible future extensions)
+   *  @param e
+   *    the comparison value for checking match
+   *  @param timed
+   *    if true, wait only until timeout elapses
+   *  @param nanos
+   *    timeout in nanosecs, used only if timed is true
+   *  @return
+   *    matched item, or e if unmatched on interrupt or timeout
+   */
+  private def awaitMatch(
+      s: Node,
+      pred: Node,
+      e: E,
+      timed: Boolean,
+      nanos: Long
+  ): E = {
+    var _nanos = nanos
+    val isData = s.isData
+    val deadline = if (timed) System.nanoTime() + _nanos else 0
+    val w = Thread.currentThread()
+    var stat = -1
+    var item: Object = s.item
+    var continueLoop = true
+    while (item == e && continueLoop) {
+      if (needSweep) sweep()
+      else if ((timed && _nanos <= 0) || w.isInterrupted()) {
+        if (s.casItem(e, if (e == null) s else null)) {
+          unsplice(pred, s) // cancelled
+          return e
+        }
+      } else if (stat <= 0) {
+        if (pred != null && pred.next == s) {
+          if (stat < 0 &&
+              (pred.isData != isData || pred.isMatched())) {
+            stat = 0 // yield once if first
+            Thread.`yield`()
+          } else {
+            stat = 1
+            s.waiter = w // enable unpark
+          }
+        } // else signal in progress
+      } else if ({ item = s.item; item != e }) {
+        continueLoop = false
+      } else if (!timed) {
+        LockSupport.setCurrentBlocker(this)
+        try {
+          ForkJoinPool.managedBlock(s)
+        } catch {
+          case cannotHappen: InterruptedException => {}
+        }
+        LockSupport.setCurrentBlocker(null)
+      } else {
+        _nanos = deadline - System.nanoTime()
+        if (_nanos > SPIN_FOR_TIMEOUT_THRESHOLD)
+          LockSupport.parkNanos(this, _nanos)
+      }
+
+      item = s.item
+    }
+    if (stat == 1)
+      s.waiterAtomic.store(null)
+    if (!isData)
+      s.itemAtomic.store(s) // self-link to avoid garbage
```

Comment:
Actually `.set` on a `VarHandle` is a plain write.

> Sets the value of a variable to the newValue, with memory semantics of setting as if the variable was declared non-volatile and non-final. Commonly referred to as plain write access. 

https://github.com/armanbilge/jsr166/blob/abd155f72243cc2066c31ee732d21894f0edf7bd/src/main/java/util/concurrent/LinkedTransferQueue.java#L650-L653

Reply:
Thanks for pointing that out! `relaxed` the store :)